### ISSUE TITLE
`tf.keras.layers.experimental.EinsumDense` gives different results depending on batch size

### ISSUE THREAD

**Issue by Ghostvv** [tensorflow/tensorflow#53321]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: no
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): 2.6.0 and 2.7.0
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

when using `tf.keras.layers.experimental.EinsumDense` layer produce different result depending on the batch size

**Describe the expected behavior**

the output of the layer should be independent of batch dimension

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
import tensorflow as tf
dense = tf.keras.layers.experimental.EinsumDense(
        "bac,acd->bda",
        output_shape=[64,4],
        bias_axes="da"
)
x = tf.random.uniform((80,4,32))
print(dense(x)[0, :4].numpy())
print(dense(x[:1])[0, :4].numpy())
```
I've got different result in 7th-8th order:
```
[[ 0.12284548  0.2814498  -0.34291047  0.00810905]
 [ 0.23635934 -0.08506497  0.12073331  0.33535597]
 [-0.30136532  0.34854767  0.41540402  0.1328382 ]
 [-0.04340287 -0.07197566  0.17427945 -0.29642397]]

[[ 0.12284552  0.2814498  -0.34291047  0.00810904]
 [ 0.23635934 -0.08506497  0.12073331  0.33535597]
 [-0.3013654   0.3485477   0.415404    0.1328382 ]
 [-0.04340289 -0.07197568  0.17427945 -0.29642397]]
```
Having several of Einsum layers leads to accumulated error and wrong predictions

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Ghostvv** [comment]
Similar behavior can be observed for the `tf.einsum`:
```
import tensorflow as tf
x = tf.random.uniform((80,4,32))
w = tf.random.uniform((4,32,64))
print(tf.einsum("bac,acd->bda", x, w)[0, :4].numpy())
print()
print(tf.einsum("bac,acd->bda", x[:1], w)[0, :4].numpy())
```
I've got different result in 7th-8th order:
```

[[8.515028  6.533536  8.1234455 6.415581 ]
 [8.387905  8.020177  8.997409  6.6566358]
 [7.485701  7.6112795 8.145813  7.514901 ]
 [8.022375  7.008651  7.9132357 5.918109 ]]

[[8.515028  6.5335355 8.123446  6.4155803]
 [8.387905  8.020177  8.997409  6.6566353]
 [7.485701  7.611279  8.145813  7.5149007]
 [8.022375  7.0086513 7.9132347 5.9181094]]
```

**Comment by Ghostvv** [comment]
The issue is rather important because attention implementation uses einsum dense layers

**Comment by sanatmpa1** [comment]
@Ghostvv,

I've tested the code in colab with `TF 2.70.0` and I am able to see the similar result for . Please find the [gist here](https://colab.sandbox.google.com/gist/sanatmpa1/6bbaead47f012b51b482722a5822c902/53321.ipynb) and let me know if I am missing something. Thanks!

**Comment by Ghostvv** [comment]
@sanatmpa1 thank you for creating colab notebook. It looks correct there, however locally I ghave `2.7.0` as well, however I get slightly different numbers:
```
>>> import tensorflow as tf
2021-12-07 09:07:10.186065: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-12-07 09:07:10.186104: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
>>> tf.__version__
'2.7.0'
>>> x = tf.random.uniform((80,4,32))
2021-12-07 09:07:41.415303: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2021-12-07 09:07:41.415326: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
2021-12-07 09:07:41.415352: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (cognigy-workstation-128): /proc/driver/nvidia/version does not exist
2021-12-07 09:07:41.415665: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
>>> w = tf.random.uniform((4,32,64))
>>> print(tf.einsum("bac,acd->bda", x, w)[0, :4].numpy())
[[6.923691  5.8655972 8.443382  6.651726 ]
 [8.643357  7.0431523 7.812662  7.4349017]
 [7.7721167 6.5131044 7.1347017 8.100251 ]
 [7.3623085 5.9433074 7.341781  6.8320265]]
>>> print()

>>> print(tf.einsum("bac,acd->bda", x[:1], w)[0, :4].numpy())
[[6.9236917 5.8655972 8.443382  6.651726 ]
 [8.643357  7.0431523 7.812662  7.4349017]
 [7.772116  6.5131044 7.1347017 8.100251 ]
 [7.3623085 5.943307  7.341781  6.8320265]]
>>> import numpy as np
>>> np.set_printoptions(12)
>>> print(tf.einsum("bac,acd->bda", x, w)[0, :4].numpy())
[[6.923691  5.8655972 8.443382  6.651726 ]
 [8.643357  7.0431523 7.812662  7.4349017]
 [7.7721167 6.5131044 7.1347017 8.100251 ]
 [7.3623085 5.9433074 7.341781  6.8320265]]
>>> print(tf.einsum("bac,acd->bda", x, w)[0, 0].numpy())
[6.923691  5.8655972 8.443382  6.651726 ]
>>> print(tf.einsum("bac,acd->bda", x, w)[0, 0, 0].numpy())
6.923691
>>> print(tf.einsum("bac,acd->bda", x[:1], w)[0, 0, 0].numpy())
6.9236917
```
Do you have an idea why? the result is that my architecture with several attentions implemented with `tf.keras.layers.MultiHeadAttention` produce different prediciton confidence for the same example depending on the prediction batch size

**Comment by Ghostvv** [comment]
also locally, I have similar problems with transformer based models from tensorflow hub:
```
import tensorflow as tf
import tensorflow_hub as hub
# Needed for loading universal-sentence-encoder-cmlm/multilingual-preprocess
import tensorflow_text  # noqa


encoder = hub.load("https://tfhub.dev/google/universal-sentence-encoder/4")


texts = [
    "schedule a booking",
    "I'd like to book an appointment, please."
    "kuhwb wuehb jweb",
    "uiweh uiwe",
    "I want to cancel my appointment"
]

print(all(encoder(texts).numpy()[-1] == encoder(texts[-1:]).numpy()[-1]))
```
I get `False`

**Comment by Ghostvv** [comment]
@sanatmpa1 I added the example with sentence encoder above and I also get different sentence embeddings depending on the batch size

**Comment by Ghostvv** [comment]
@sanatmpa1 sorry for multiple messages 🙈 I increased the seizes of the vectors and got different results in your collar notebook for `tf.einsum` as well

**Comment by weeklynote** [comment]
已收到，我会尽快回复你！
                     ——刘劲字

**Comment by cuixue** [comment]
您好，邮件已经收到，我会尽快处理的。谢谢

**Comment by sachinprasadhs** [comment]
e-7 or higher floating point precision error is expected and are negligible which does not impact much on the overall precision/accuracy.

**Comment by Ghostvv** [comment]
I have attention intent model and I noticed that the same sentence has different intent prediction depending on the position in the batch. My guess would be that this error got accumulated given many layers

**Comment by sachinprasadhs** [comment]
Note: This level of error is at the 1 ulp (unit in the last place) for float32. This is expected as the order of operations in the algorithm may differ depending on batch size, due to different traversal order or parallelization strategy. Please do not ever rely on exact floating point accuracy. Thanks!
